libs/utils/graph/iso.rs

```diff
@@ -23,7 +54,7 @@ fn max_matching<EK: EdmondsKarp<i32>>(
     graph_g: &Graph<(String, IndexMap<String, String>), String>,
     graph_h: &Graph<(String, IndexMap<String, String>), String>,
     set_s: &IndexMap<
-        (NodeIndex, NodeIndex),
```

Comment:
What kind of data structure is NodeIndex anyway? Do we only use it to get the index from it?

Reply:
Honestly, it might be easier to just use a typedef for this

Reply:
NodeIndex is something internal to Petgraph.  They use it as a wrapper around the indices for nodes.  We use it as a handle to identify nodes in the graph.  Here, I got the index, because the NodeIndex object serializes as, for example, "NodeIndex(5)".  Except we know it's a NodeIndex, and don't need the extra information identifying it.  It is both shorter and easier to parse for the deserializer if we get rid of the NodeIndex label by calling .index().

Reply:
So at what point do we need to use NodeIndex. Can we just carry the int around?

Reply:
We need to use NodeIndex when we're actually performing the isomorphism algorithm.  We also use NodeIndex when we access nodes in the target or trace graphs.  What do you mean by carry the int around?  Parse everything to a NodeIndex within the isomorphism code?

Reply:
So apparently we only really need an integer index? Or does NodeIndex maintain some refs. So we could just convert the integer to a NodeIndex object once we actually need it for petgraph. Otherwise, we serialize and deserialize int values.

Reply:
Yeah, that could work.  The thing is, the values are never useful as ints.  You can't plug an int into anything in petgraph and do stuff.  So if those values are ever to be used, they have to be NodeIndexs.  We might be able to save something in that we may not always use all entries in set S at a given node, but I'm not sure we'd save much.

Reply:
Well I was just wondering. Integers are trivially serializable and the only purpose is to use it for Petgraph. So why not (de)serialize integers instead of custom data structures? Then once you need a NodeIndex for Petgraph, create it from scratch. Instead of Vec<(NodeIndex, NodeIndex) you will have Vec<(int, int)>. What information does a NodeIndex actually store?

Reply:
NodeIndex is just a wrapper for the Ix type, which is itself just an alias for usize (https://docs.rs/ndarray/0.12.0/ndarray/type.Ix.html).  The issue for serialization isn't with NodeIndex as a custom type, the issue is that tuples of any kind do not serialize as keys (https://github.com/serde-rs/json/issues/456).  So if we had a tuple of ints, we'd still need to do custom (de)serialization.

Reply:
Okay will approve
But why does #[derive(Debug, Hash, Eq, PartialEq, Serialize, Deserialize)] for the struct not work? Is that an issue during runtime? It seems to compile fine at least.

Reply:
It's an issue during runtime, yeah.  If you do a small example where you serialize and then subsequently deserialize, it will error out, saying that keys must be strings.

Reply:
I've got a small example in rust playground showing the error - I think I can link it with this gist, but haven't tried before so it may not work.

https://gist.github.com/9ac05deaad824cd8588075585167caea

Reply:
Oh I understand now, thanks for the gist. I think this can easily be solved by using simpler data structures, using a tuple as hash map key is fairly awkward. In general, a data structure like `IndexMap<SetSKey, IndexMap<NodeIndex, Option<Vec<(NodeIndex, NodeIndex)>>>` raises a bunch of flags. This can probably simplified.